### PR TITLE
join fixups

### PR DESCRIPTION
**Pull request by mmaxim**

### CODE REVIEW

File: shared/chat/conversation/messages/system-joined/container.tsx [hunk 1]
```diff
@@ -12,7 +12,7 @@ const mapStateToProps = (state, {message}) => ({
   _meta: Constants.getMeta(state, message.conversationIDKey),
   author: message.author,
   authorIsYou: state.config.username === message.author,
-  joiners: message.joiners,
+  joiners: !message.joiners.length && !message.leavers.length ? [message.author] : message.joiners,
```
Comment (joshblum) [inline shared/chat/conversation/messages/system-joined/container.tsx]: why is this necessary? why doesn't the service do this?

Comment (mmaxim) [reply shared/chat/conversation/messages/system-joined/container.tsx]: The problem is if you are looking at the thread, it doesn't go through the whole consolidation path.

**Comment by mmaxim** [comment]
Right now the join/leave/join is like "mikem, mikem joined. mikem left". Is that anymore clear?

**Comment by joshblum** [comment]
no the status quo isn't great either

File: shared/chat/conversation/messages/system-joined/container.tsx (hunk 1 above)
Comment (joshblum) [reply shared/chat/conversation/messages/system-joined/container.tsx]: oh so this was preexisting?

**Comment by mmaxim** [comment]
I guess we can not consolidate in that case

File: go/chat/uithreadloader.go [hunk 2]
```diff
@@ -76,11 +77,25 @@ func (t *UIThreadLoader) groupThreadView(ctx context.Context, uid gregor1.UID, t
...
+			if err != nil {
+				return false
+			}
+			if !(mtyp == chat1.MessageType_JOIN || mtyp == chat1.MessageType_LEAVE) {
+				return false
+			}
+			if msg.Valid().ClientHeader.Sender.Eq(uid) {
+				return false
+			}
+			for _, g := range grouped {
```
Comment (joshblum) [inline go/chat/uithreadloader.go]: might be better to have a global join/leave map so we don't have to traverse the loop repeatedly

Comment (mmaxim) [reply go/chat/uithreadloader.go]: This is the current set of consolidated messages. It won't be long.

File: go/chat/uithreadloader.go [hunk 3]
```diff
@@ -59,14 +59,16 @@ func (t *UIThreadLoader) groupGeneric(ctx context.Context, uid gregor1.UID, msgs
...
-				grouped = append(grouped, msg)
-				continue
-			}
+		if msg.IsValid() && matches(msg, grouped) {
+			grouped = append(grouped, msg)
+			continue
 		}
 		addGrouped()
-		res = append(res, msg)
+		if msg.IsValid() && matches(msg, grouped) {
```
Comment (joshblum) [inline go/chat/uithreadloader.go]: worthwhile to leave a comment as to why this is necessary
